### PR TITLE
Fix original target framework normalization in XPLAT case

### LINKED ISSUE #6105: [Regression] in 15.5p1: Portable0.0 is skipped
_From @onovotny on October 18, 2017 23:57_

I have an SDK project that the `MSBuild.Sdk.Extras` to build a legacy PCL target. This works in 15.4 but fails in 15.5 preview 1.

In the IDE in 15.4, I see this:

image

In 15.5 p1 I see this:

image
Note that the `Portable0.0` entry representing `portable-net4+sl5+wp8+win8+wpa81` is missing.

This is a blocker.

Building in the IDE in 15.5 p1 also yields the following error:
```
Severity	Code	Description	Project	File	Line	Suppression State
Error		The TargetFramework value 'portable-net4+sl5+wp8+win8+wpa81' was not recognized. It may be misspelled. If not, then the TargetFrameworkIdentifier and/or TargetFrameworkVersion properties must be specified explicitly.	crypto	C:\Program Files\dotnet\sdk\2.0.2\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.TargetFrameworkInference.targets	96	
```
This is incorrect as the TFI and TFV is set by the MSBuild.Sdk.Extras package.

Repro is https://github.com/onovotny/bc-csharp/tree/e55eb138bf5492769ec1cf2212b49f223e377145
Open the BouncyCastle.sln

/cc @nguerrera 

_Copied from original issue: dotnet/sdk#1666_

_Copied from original issue: dotnet/project-system#2910_

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug
Fixes: https://github.com/NuGet/Home/issues/6105
Regression: Yes, Introduced in 15.5 before preview 1. 
If Regression then when did it last work: 15.4
If Regression then how are we preventing it in future:  Adding more automated tests. 

## Fix
Details: 
The regression was introduced in https://github.com/NuGet/NuGet.Client/pull/1659.

The reason for that was to make No-Op work xplat, as there were differences between VS & msbuild/dotnet restore. 

The original target frameworks value is used in the nuget.g.targets and props to condition based on the TFM. 
This needs to be the same as the user written value otherwise these targets won't get imported, because build uses the exact value user provided in their <TargetFrameworks> property. 

The fix here is to not normalize in the crossTargeting case (TargetFramework**s**). 
The normalization in the crossTargeting=false case needs to be there for no-op consistency. It does not have any other impact.  

## Testing/Validation
Tests Added: Yes
Reason for not adding tests:  
Validation done:  Manual + Automated tests

PS. 

Created a 4.5.0-rtm branch as well. 
Fixed the private branch tests, and running those tests as well.

https://devdiv.visualstudio.com/DevDiv/_build/index?buildId=1095865&_a=summary

Once the tests are done, I will drop an e-mail to the vendors so they can test it and get an RPS run.

### CODE REVIEW

File: src/NuGet.Core/NuGet.ProjectModel/PackageSpecWriter.cs [hunk 1]
```diff
@@ -160,7 +160,14 @@ private static void SetMSBuildMetadata(IObjectWriter writer, PackageSpec package
 
             SetArrayValue(writer, "fallbackFolders", msbuildMetadata.FallbackFolders);
             SetArrayValue(writer, "configFilePaths", msbuildMetadata.ConfigFilePaths);
-            SetArrayValue(writer, "originalTargetFrameworks", msbuildMetadata.OriginalTargetFrameworks.Select(e => NuGetFramework.Parse(e).GetShortFolderName()));
+            if (msbuildMetadata.CrossTargeting)
+            {
+                SetArrayValue(writer, "originalTargetFrameworks", msbuildMetadata.OriginalTargetFrameworks.OrderBy( c => c)); // This need to stay the original strings because the nuget.g.targets have conditional imports based on the original framework name
```
Comment (emgarten) [inline src/NuGet.Core/NuGet.ProjectModel/PackageSpecWriter.cs]: Add Ordinal compare for sorting

File: src/NuGet.Core/NuGet.ProjectModel/PackageSpecWriter.cs [hunk 2]
```diff
@@ -160,7 +160,14 @@ private static void SetMSBuildMetadata(IObjectWriter writer, PackageSpec package
...
+            }
+            else
+            {
+                SetArrayValue(writer, "originalTargetFrameworks", msbuildMetadata.OriginalTargetFrameworks.Select(e => NuGetFramework.Parse(e).GetShortFolderName()).OrderBy(c => c));
```
Comment (emgarten) [inline src/NuGet.Core/NuGet.ProjectModel/PackageSpecWriter.cs]: Ordinal here also

**Review by jainaashish — APPROVED** [review]
Looks good! 👏

File: test/TestUtilities/Test.Utility/SimpleTestSetup/SimpleTestProjectContext.cs [hunk 3]
```diff
@@ -214,7 +214,8 @@ public PackageSpec PackageSpec
                     _packageSpec.RestoreMetadata.ProjectPath = ProjectPath;
                     _packageSpec.RestoreMetadata.ProjectStyle = Type;
                     _packageSpec.RestoreMetadata.OutputPath = AssetsFileOutputPath;
-                    _packageSpec.RestoreMetadata.OriginalTargetFrameworks = OriginalFrameworkStrings;
+                    _packageSpec.RestoreMetadata.OriginalTargetFrameworks = OriginalFrameworkStrings
+                        ;
```
Comment (emgarten) [inline test/TestUtilities/Test.Utility/SimpleTestSetup/SimpleTestProjectContext.cs]: nit: formatting

File: test/TestUtilities/Test.Utility/SimpleTestSetup/SimpleTestProjectContext.cs [hunk 4]
```diff
@@ -411,7 +412,8 @@ public XDocument GetXML()
 
                     ProjectFileUtils.AddProperties(xml, new Dictionary<string, string>()
                     {
-                        { tfPropName, string.Join(";", Frameworks.Select(f => f.Framework.GetShortFolderName())) },
+                        { tfPropName, OriginalFrameworkStrings.Count != 0 ? string.Join(";", OriginalFrameworkStrings):
```
Comment (emgarten) [inline test/TestUtilities/Test.Utility/SimpleTestSetup/SimpleTestProjectContext.cs:414]: 🥇

**Review by emgarten — APPROVED** [review]
🚀

**Pull request merged**